rfd/0057-automatic-aws-server-discovery.md

```diff
@@ -0,0 +1,289 @@
+will require the teleport SSH agent to include `ec2:DescribeInstances` as part of
+it's IAM permissions
+
+As with AWS database discover, new EC2 nodes will be discovered periodically on a 60
+second timer, as new nodes are found they will be added to the teleport cluster.
+
+In order to avoid attempting to reinstall teleport on top of an instance where it is
+already present the generated teleport config will include a static label indicating
+it was created via auto discovery.
+
+Example:
+```json
+{
+  "kind": "node",
+  "version": "v2",
+  "metadata": {
+    "name": "AWS_INSTANCE_ID",
+    "labels": {
+      "env": "example"
+      "teleport.dev/discovered-node": "yes"
+    },
+  },
+  "spec": {
+    "public_addr": "ec2-54-194-252-215.us-west-1.compute.amazonaws.com",
+    "hostname": "awsxyz"
+  }
+}
+```
+
+IAM join tokens will have a `tags` field added to the accounts so so only EC2
+instances with matching `tags` can be added using the token. the tags matching will
+default to `"*":"*"` to retain compatibility with previously generated IAM tokens
+
+## Agent installation
+
+In order to install the Teleport agent on EC2 instances, the user must create a
+custom SSM Command document. By using the document it allows for only a specific set
+of install commands to be run, rather than having a whole shell available.
+
+Example SSM aws:runCommand document:
+```yaml
+# name: installTeleport
+---
+schemaVersion: '2.2'
+description: aws:runShellScript
+mainSteps:
+- action: aws:runShellScript
+  name: runShellScript
+  inputs:
+    timeoutSeconds: '300'
+    runCommand:
+      - "#!/bin/sh"
+      - curl https://.../teleport-pubkey.asc ...
+      - echo "deb [signed-by=... stable main" | tee ... > /dev/null
+      - apt-get update
+      - apt-get install teleport
+      - teleport ssh configure --auth-agent=...  --join-method=iam --token-name=iam-
+
+```
+
+In order to run the new SSM document the AWS user will need IAM permissions to run
+SSM commands[3] for example:
+
+```json
+{
+    "Statement": [
+        {
+            "Action": "ssm:SendCommand",
+            "Effect": "Allow",
+            "Resource": [
+                # Allow running commands on all us-west-2 instances
+                "arn:aws:ssm:us-west-2:*:instance/*",
+                 # Allows running the installTeleport docuemnt on the allowed instances
+                "arn:aws:ssm:us-east-2:aws-account-ID:document/installTeleport"
+            ]
+        }
+    ]
+}
+```
+
+On AWS, Amazon Linux and Ubuntu LTS (16.04, 18.04, 20.04) come with the SSM agent
+preinstalled[4].
+
+## Config generation
+
+A new `teleport` subcommand will also be added -- `teleport ssh configure`
+
+This will be used to generate a new /etc/teleport.yaml file and support the following arguments:
+```
+teleport ssh configure
+    --auth-server=auth-server.example.com [auth server that is being connected to]
+    --join-method=iam
+    --token-name=iam-token-name
+    --auto-discovered
```

Comment:
may be use present time instead of past here? `--auto-discover`? I don't think we use past time anywhere else in the code/tools

Reply:
If this flag is needed to set the "origin" label, then we can just use the existing `--labels` flag.